hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java

```diff
@@ -60,158 +67,146 @@ public TransactionDispatcher(
-            };
-
-            case NODE_STAKE_UPDATE, DATA_NOT_SET -> throw new UnsupportedOperationException(
-                    "Not implemented");
+                    .preHandle(handlerContext, setupTokenStore(statesTracker));
+            case TOKEN_PAUSE -> handlers.tokenPauseHandler().preHandle(handlerContext);
+            case TOKEN_UNPAUSE -> handlers.tokenUnpauseHandler().preHandle(handlerContext);
+
+            case UTIL_PRNG -> handlers.utilPrngHandler().preHandle(handlerContext);
+
+            case SYSTEMDELETE -> {
+                switch (txBody.getSystemDelete().getIdCase()) {
+                    case CONTRACTID -> handlers.contractSystemDeleteHandler()
+                            .preHandle(handlerContext);
+                    case FILEID -> handlers.fileSystemDeleteHandler().preHandle(handlerContext);
+                    case ID_NOT_SET -> throw new IllegalArgumentException(
+                            "SystemDelete without IdCase");
+                }
+            }
+            case SYSTEMUNDELETE -> {
+                switch (txBody.getSystemUndelete().getIdCase()) {
+                    case CONTRACTID -> handlers.contractSystemUndeleteHandler()
+                            .preHandle(handlerContext);
+                    case FILEID -> handlers.fileSystemUndeleteHandler().preHandle(handlerContext);
+                    case ID_NOT_SET -> throw new IllegalArgumentException(
+                            "SystemUndelete without IdCase");
+                }
+            }
+
+            default -> throw new UnsupportedOperationException(TYPE_NOT_SUPPORTED);
+        }
+    }
+
+    private static ReadableAccountStore setupAccountStore(
+            @NonNull final ReadableStatesTracker statesTracker) {
+        final var tokenStates = statesTracker.getReadableStates(TOKEN_SERVICE_KEY);
+        return new ReadableAccountStore(tokenStates);
+    }
+
+    private static ReadableScheduleStore setupScheduleStore(
+            @NonNull final ReadableStatesTracker statesTracker) {
+        final var scheduleStates = statesTracker.getReadableStates(SCHEDULE_SERVICE_KEY);
+        return new ReadableScheduleStore(scheduleStates);
+    }
+
+    private ReadableTokenStore setupTokenStore(@NonNull final ReadableStatesTracker statesTracker) {
+        final var tokenStates = statesTracker.getReadableStates(TOKEN_SERVICE_KEY);
+        return new ReadableTokenStore(tokenStates);
+    }
+
+    private PreHandleDispatcher setupPreHandleDispatcher(
+            @NonNull final ReadableStatesTracker statesTracker) {
+        return (TransactionBody innerTxn, AccountID innerPayer) -> {
+            final var accountStore = setupAccountStore(statesTracker);
+            final var handlerContext =
+                    new PrehandleHandlerContext(accountStore, innerTxn, innerPayer);
+            dispatchPreHandle(statesTracker, handlerContext);
+            return new TransactionMetadata(handlerContext, List.of());
         };
```

Comment:
Should we probably move these methods to a different class , since we will get more stores ?

Reply:
Good idea